src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+        // Determine if it needs to unwind or wind
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+        uint256 Line;
+        uint256 debt;
+        uint256 targetAssets;
+        uint256 toUnwind;
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+        // Determine if it needs to fully unwind due to D3M ilk being caged (but not culled) or plan is not active
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) {
+            toUnwind = type(uint256).max; // We make sure to enter the unwind path
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+            Line = vat.Line();
+            debt = vat.debt();
+            targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
 
-        emit Unwind(ilk, amount, fees);
-    }
+            // Determine if it needs to unwind due to:
+            unchecked {
+                toUnwind = _max(
+                                _max(
+                                    art > lineWad ? art - lineWad : 0, // ilk debt ceiling exceeded
+                                    debt > Line ? _divup(debt - Line, RAY) : 0 // global debt ceiling exceeded
+                                ),
+                                targetAssets < currentAssets ? currentAssets - targetAssets : 0 // plan targetAssets
+                            );
+            }
+        }
 
-    function _fix(bytes32 ilk, address _pool) internal returns (uint256 diff) {
-        (uint256 ink, uint256 art) = vat.urns(ilk, _pool);
-        if (art < ink) {
-            address _vow = vow;
-            diff = ink - art;
-            require(diff <= MAXINT256, "D3MHub/overflow");
-            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
-            vat.grab(ilk, _pool, _pool, _vow, 0, int256(diff));
+        if (toUnwind > 0) {
+            toUnwind = _min(
+                toUnwind,
+                maxWithdraw
+            );
+            require(toUnwind <= MAXINT256, "D3MHub/overflow");
```

Comment:
This can get stuck in the extreme edge case where total `vat.debt` is in excess of `MAXINT256`, the global `Line` gets set to `0`, and `maxWithdraw` is reported in excess of `MAXINT256` by an integration.

More generally, for whatever reason, in a case where both `toUnwind` and `maxWithdraw` are in excess of `MAXINT256` we can prevent revert by simply adding `MAXINT256` to the `_min()` condition.  This would prevent all cases where this contract is bricked because of the `require()`.

NOTE: If others agree with this change, we may want to consider a similar change in the `_wind()` condition below.
```suggestion
            toUnwind = _min(
                _min(
                    toUnwind,
                    maxWithdraw
                ),
                MAXINT256 
            );
```

Reply:
We need to be careful, and think it better, as the rebalance counts to unwinding the full amount, if we do partial unwinding we can break the rule there.
Before this case wasn't actually possible to reach that value, as winding was always limited to `Art + toWind <= maxint256`, so we couldn't end with a CDP higher than that number.
However now with the rebalance that might happen. What probably is better is limiting the rebalance also to end up `<= maxint256`. Otherwise thinking that there will be that amount of DAI to redeem is because all is already broken.

Reply:
This thing should be addressed now.